plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Zip.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+module UntypedPlutusCore.Core.Zip
+    ( pzipWith
+    , pzip
+    , tzipWith
+    , tzip
+    ) where
+
+import Control.Monad (when)
+import PlutusPrelude
+import UntypedPlutusCore.Core.Instance.Eq ()
+import UntypedPlutusCore.Core.Type
+
+-- | Zip two programs using a combinator function for annotations.
+-- Fail if the input programs are not "equal" modulo annotations.
+pzipWith :: forall p name uni fun ann1 ann2 ann3 m.
+           (p ~ Program name uni fun, (Eq (Term name uni fun ())), MonadFail m)
```

Comment:
Can we just never use `MonadFail`. IMO even `MonadError String` is better.

Reply:
good call